### PR TITLE
MG-2053 : Structure API Endpoints Based on Entities

### LINKED ISSUE #2053: Feature: Enhance OpenAPI Specification by Structuring Endpoints Based on Entities for Improved Clarity and User Experience
### Is your feature request related to a problem? Please describe.

Structuring endpoints based on entities rather than service-specific endpoints, the proposal seeks to streamline the understanding of functionalities related to specific entities, such as users, things, groups, channels, domains making it easier for users to navigate and comprehend the available API functionalities within a specific context.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Structuring API endpoints not based on service-specific endpoints but rather on entities, For example, within the `users.yml` file, all endpoints would start with /users. Operations related to specific users, such as retrieving user tags (`users/{userID}/tag`), would be included here. However, I propose including additional endpoints, like `/users/{userID}/things`, which pertain to the `things` associated with a user, even if the `things` functionality resides in a different service.

By including such endpoints in the `users.yml` API specification, users interacting with this document will find it easier to understand as they would expect `users.yml` to provide comprehensive user-related functionalities, even if certain functionalities are fulfilled by other services.

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by WashingtonKK**
# What type of PR is this?

This is a refactor because it changes the naming and structuring of API endpoints.

## What does this do?

This restructures API endpoints so that they are named based on entities such as groups, channels, users, things etc.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2053
- Resolves #2053

## Have you included tests for your changes?

No, I have not included tests.

## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

### CODE REVIEW

File: api/openapi/auth.yml [hunk 1]
```diff
@@ -85,7 +85,7 @@ paths:
         "500":
           $ref: "#/components/responses/ServiceError"
 
-  /domains/{domainID}:
+  /domains/{domainID}/domain:
```
Comment (rodneyosodo) [inline api/openapi/auth.yml]: No need for this. We can revert `/domains/{domainID}` is elaborate

File: api/openapi/auth.yml [hunk 2]
```diff
@@ -302,6 +302,39 @@ paths:
           description: Database can't process request.
         "500":
           $ref: "#/components/responses/ServiceError"
+  /domains/users/{memberID}:
```
Comment (rodneyosodo) [inline api/openapi/auth.yml]: Also update SDK

File: users/api/groups.go [hunk 3]
```diff
@@ -125,23 +125,22 @@ func groupsHandler(svc groups.Service, r *chi.Mux, logger *slog.Logger) http.Han
 			api.EncodeResponse,
 			opts...,
 		), "unassign_groups").ServeHTTP)
+
+		r.Get("/channels/{memberID}", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (rodneyosodo) [inline users/api/groups.go]: Update SDK

File: users/api/clients.go [hunk 4]
```diff
@@ -124,6 +124,34 @@ func clientsHandler(svc users.Service, r *chi.Mux, logger *slog.Logger, pr *rege
 			api.EncodeResponse,
 			opts...,
 		), "disable_client").ServeHTTP)
+
+		r.Get("/groups/{groupID}", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (rodneyosodo) [inline users/api/clients.go]: Also update SDK

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/domains.go [hunk 5]
```diff
@@ -122,7 +122,7 @@ func (sdk mgSDK) Domains(pm PageMetadata, token string) (DomainsPage, errors.SDK
 }
 
 func (sdk mgSDK) ListDomainUsers(domainID string, pm PageMetadata, token string) (UsersPage, errors.SDKError) {
-	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s/%s", domainsEndpoint, domainID, usersEndpoint), pm)
+	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s/%s", usersEndpoint, domainsEndpoint, domainID), pm)
```
Comment (rodneyosodo) [inline pkg/sdk/go/domains.go]:
```suggestion
	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s/%s", domainsEndpoint, usersEndpoint, domainID), pm)
```

File: pkg/sdk/go/channels.go [hunk 6]
```diff
@@ -167,7 +167,7 @@ func (sdk mgSDK) RemoveUserFromChannel(channelID string, req UsersRelationReques
 }
 
 func (sdk mgSDK) ListChannelUsers(channelID string, pm PageMetadata, token string) (UsersPage, errors.SDKError) {
```
Comment (rodneyosodo) [inline pkg/sdk/go/channels.go]: This should go to users file

File: pkg/sdk/go/channels.go [hunk 7]
```diff
@@ -208,7 +208,7 @@ func (sdk mgSDK) RemoveUserGroupFromChannel(channelID string, req UserGroupsRequ
 }
 
 func (sdk mgSDK) ListChannelUserGroups(channelID string, pm PageMetadata, token string) (GroupsPage, errors.SDKError) {
```
Comment (rodneyosodo) [inline pkg/sdk/go/channels.go]: This should go to groups file

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Some endpoints have not been updated on the SDK for example
- `channels/users/userID`
- `channesl/groups/groupID`
- `groups/channels/channelID`
- `users/domains/domainID`

File: users/api/clients.go [hunk 8]
```diff
@@ -124,6 +124,34 @@ func clientsHandler(svc users.Service, r *chi.Mux, logger *slog.Logger, pr *rege
...
+		), "list_users_by_user_group_id").ServeHTTP)
+
+		r.Get("/channels/{channelID}", otelhttp.NewHandler(kithttp.NewServer(
+			listMembersByChannelEndpoint(svc),
+			decodeListMembersByChannel,
+			api.EncodeResponse,
+			opts...,
+		), "list_users_by_channel_id").ServeHTTP)
+
+		r.Get("/things/{thingID}", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (rodneyosodo) [inline users/api/clients.go]: This endpoint is not documented

File: users/api/clients.go [hunk 9]
```diff
@@ -124,6 +124,34 @@ func clientsHandler(svc users.Service, r *chi.Mux, logger *slog.Logger, pr *rege
...
+			listMembersByThingEndpoint(svc),
+			decodeListMembersByThing,
+			api.EncodeResponse,
+			opts...,
+		), "list_users_by_thing_id").ServeHTTP)
+
+		r.Get("/domains/{domainID}", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (rodneyosodo) [inline users/api/clients.go]: This endpoint is not documented

File: things/api/http/clients.go [hunk 10]
```diff
@@ -115,26 +115,22 @@ func clientsHandler(svc things.Service, r *chi.Mux, logger *slog.Logger) http.Ha
...
 		), "delete_thing").ServeHTTP)
+
+		r.Get("/channels/{groupID}/", otelhttp.NewHandler(kithttp.NewServer(
+			listMembersEndpoint(svc),
+			decodeListMembersRequest,
+			api.EncodeResponse,
+			opts...,
+		), "list_things_by_channel_id").ServeHTTP)
+
+		r.Get("/users/{userID}/", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (rodneyosodo) [inline things/api/http/clients.go]: This endpoint is not documented

File: pkg/sdk/go/groups.go [hunk 11]
```diff
@@ -187,7 +187,7 @@ func (sdk mgSDK) RemoveUserFromGroup(groupID string, req UsersRelationRequest, t
 }
 
 func (sdk mgSDK) ListGroupUsers(groupID string, pm PageMetadata, token string) (UsersPage, errors.SDKError) {
```
Comment (rodneyosodo) [inline pkg/sdk/go/groups.go]: This should be in `users.go`. Cross check other SDK:
- for example `/things/channels/channelID` should be in things.go
- `/channels/users/userID` should be in `channels.go`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: users/api/clients.go (hunk 9 above)
Comment (WashingtonKK) [reply users/api/clients.go]: It is documented here:
https://github.com/WashingtonKK/magistrala/blob/69b579199035c44acc352e3f588f0eba2ddac163/api/openapi/users.yml#L974-L975

File: pkg/sdk/go/groups.go (hunk 11 above)
Comment (WashingtonKK) [reply pkg/sdk/go/groups.go]: There is no `/things/channels/channelID` endpoint. We have `/things/channels/thingID` which lists channels by thingID, and I think this one should remain in the channels.go file.

What do you think @rodneyosodo ?

Comment (rodneyosodo) [reply pkg/sdk/go/groups.go]: There is https://github.com/WashingtonKK/magistrala/blob/69b579199035c44acc352e3f588f0eba2ddac163/things/api/http/clients.go#L119

File: things/api/http/endpoints_test.go [hunk 12]
```diff
@@ -1919,11 +1919,10 @@ func TestListMembers(t *testing.T) {
 		req := testRequest{
 			client:      ts.Client(),
 			method:      http.MethodGet,
-			url:         ts.URL + fmt.Sprintf("/channels/%s/things?", tc.groupdID) + tc.query,
+			url:         ts.URL + fmt.Sprintf("/things/channels/%s/?", tc.groupdID) + tc.query,
```
Comment (rodneyosodo) [inline things/api/http/endpoints_test.go]: also change test for `/things/users/ID`

File: api/openapi/users.yml [hunk 13]
```diff
@@ -1002,6 +1002,37 @@ paths:
           description: Database can't process request.
         "500":
           $ref: "#/components/responses/ServiceError"
+  /users/things/{memberID}:
+    get:
+      summary: List users by thing
```
Comment (rodneyosodo) [inline api/openapi/users.yml]: reword

File: pkg/sdk/go/users.go [hunk 14]
```diff
@@ -269,26 +269,25 @@ func (sdk mgSDK) UpdateUserRole(user User, token string) (User, errors.SDKError)
 	return user, nil
 }
 
-func (sdk mgSDK) ListUserChannels(userID string, pm PageMetadata, token string) (ChannelsPage, errors.SDKError) {
-	url, err := sdk.withQueryParams(sdk.thingsURL, fmt.Sprintf("%s/%s/%s", usersEndpoint, userID, channelsEndpoint), pm)
+func (sdk mgSDK) ListChannelUsers(channelID string, pm PageMetadata, token string) (UsersPage, errors.SDKError) {
+	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s/%s", channelsEndpoint, usersEndpoint, channelID), pm)
```
Comment (rodneyosodo) [inline pkg/sdk/go/users.go]: users URL start with either `/users` or `/groups` This is not corrrect

File: pkg/sdk/go/channels.go [hunk 15]
```diff
@@ -89,6 +89,42 @@ func (sdk mgSDK) ChannelsByThing(thingID string, pm PageMetadata, token string)
 	return cp, nil
 }
 
+func (sdk mgSDK) ListGroupChannels(groupID string, pm PageMetadata, token string) (GroupsPage, errors.SDKError) {
+	url, err := sdk.withQueryParams(sdk.thingsURL, fmt.Sprintf("%s/%s/%s", groupsEndpoint, channelsEndpoint, groupID), pm)
```
Comment (rodneyosodo) [inline pkg/sdk/go/channels.go]: same

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
For SDK changes add tests to make sure they work

File: pkg/sdk/go/things.go [hunk 16]
```diff
@@ -97,8 +97,43 @@ func (sdk mgSDK) Things(pm PageMetadata, token string) (ThingsPage, errors.SDKEr
...
+	}
+	cp := ChannelsPage{}
+	if err := json.Unmarshal(body, &cp); err != nil {
+		return ChannelsPage{}, errors.NewSDKError(err)
+	}
+
+	return cp, nil
+}
+
+func (sdk mgSDK) ListUserThings(userID string, pm PageMetadata, token string) (ThingsPage, errors.SDKError) {
```
Comment (rodneyosodo) [inline pkg/sdk/go/things.go]: Add test for the other functions that have been added during this PR. You can use the following to check untested cases

```go
go test -tags test -v --race -covermode=atomic -coverprofile=cover.out ./... && go tool cover -html cover.out -o cover.html
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by felixgateru — CHANGES_REQUESTED** [review]
Consider updating the `sdk.go` with updated documentation for the new sdk functions. Also you can regenerate the mocks after making changes to the interface.

**Comment by WashingtonKK** [comment]
> Consider updating the `sdk.go` with updated documentation for the new sdk functions. Also you can regenerate the mocks after making changes to the interface.

It is already up to date @1998-felix

**Review by felixgateru — DISMISSED** [review]
LGTM

**Review by rodneyosodo — APPROVED** [review]
LGTM

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Regarding the conversation that was held change this to use query parameters for example

`/channels/things/thingID` to `/channels?thing=<id>`

**Review by arvindh123 — CHANGES_REQUESTED** [review]
At present we have `/things/thingID/channels` to list channels
After this PR https://github.com/absmach/magistrala/pull/2195  it will be like
`/channels/things/thingID` to list channels
which doesn't looks good and ideal

I have discussed this with @dborovcanin , He proposed like this  `/channels?thing=<id>`

I agree with his proposal,

Already we are having `/channels` which is used for listing all channels,
 `/channels?thing=<id>`  can be used to listing channels of particular thing id.
 `/channels?user=<id>`  can be used to listing channels of particular user id.
and so on.

File: internal/api/common.go [hunk 17]
```diff
@@ -58,6 +58,11 @@ const (
 	SharedVisibility = "shared"
 	MyVisibility     = "mine"
 	AllVisibility    = "all"
+	ThingKey         = "things"
+	UserKey          = "users"
+	ChannelKey       = "channel"
+	ChannelsKey      = "channels"
+	DomainKey        = "domains"
 	// ContentType represents JSON content type.
```
Comment (arvindh123) [inline internal/api/common.go]:
```suggestion
	ThingKey         = "thing"
	UserKey          = "user"
	ChannelKey       = "channel"
	DomainKey        = "domain"
```

File: users/api/clients.go [hunk 18]
```diff
@@ -418,58 +386,63 @@ func decodeChangeClientStatus(_ context.Context, r *http.Request) (interface{},
...
-		token:    apiutil.ExtractBearerToken(r),
-		Page:     page,
-		objectID: chi.URLParam(r, "domainID"),
+		token:      apiutil.ExtractBearerToken(r),
+		Page:       page,
+		objectID:   entityID,
+		objectKind: entityType,
 	}
 
 	return req, nil
```
Comment (arvindh123) [inline users/api/clients.go]:
```suggestion
var queryKeys = []string{api.ThingKey, api.ChannelKey, api.DomainKey, api.GroupKey, api.UserKey}
var entityTypes = map[string]string{
	api.ThingKey:   auth.ThingsKind,
	api.ChannelKey: auth.GroupsKind,
	api.DomainKey:  auth.DomainsKind,
	api.GroupKey:   auth.GroupsKind,
	api.UserKey:    auth.UsersKind,
}

func decodeListMembers(_ context.Context, r *http.Request) (interface{}, error) {
	page, err := queryPageParams(r, api.DefPermission)
	if err != nil {
		return nil, err
	}

	var entityID, entityType string

	for _, key := range queryKeys {
		entityID, err = apiutil.ReadStringQuery(r, key, "")
		if err != nil {
			return nil, errors.Wrap(apiutil.ErrValidation, err)
		}
		if entityID != "" {
			entityType = entityTypes[key]
			break
		}
	}

	req := listMembersByObjectReq{
		token:      apiutil.ExtractBearerToken(r),
		Page:       page,
		objectID:   entityID,
		objectKind: entityType,
	}

	return req, nil
```

File: internal/groups/api/decode.go [hunk 19]
```diff
@@ -56,11 +71,32 @@ func DecodeListGroupsRequest(_ context.Context, r *http.Request) (interface{}, e
...
+		memberKind = "channels"
+		entityID = ch
+	case user != "":
+		memberKind = api.UserKey
+		entityID = user
+	default:
+		entityID = ""
+		memberKind = "groups"
+	}
+
```
Comment (arvindh123) [inline internal/groups/api/decode.go]: Please fix this also like suggestion shown in users

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: api/openapi/things.yml [hunk 20]
```diff
@@ -431,6 +431,36 @@ paths:
...
+          description: Missing or invalid access token provided.
+        "403":
+          description: Failed to perform authorization over the entity.
+        "404":
+          description: A non-existent entity request.
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
```
Comment (felixgateru) [inline api/openapi/things.yml]: Please update the documentation for the changes

File: api/openapi/things.yml [hunk 21]
```diff
@@ -962,6 +963,35 @@ paths:
...
+          description: Missing or invalid access token provided.
+        "403":
+          description: Failed to perform authorization over the entity.
+        "404":
+          description: Thing does not exist.
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
```
Comment (felixgateru) [inline api/openapi/things.yml]: same

File: api/openapi/users.yml [hunk 22]
```diff
@@ -371,6 +371,44 @@ paths:
...
+            This endpoint is available only for administrators.
+        "403":
+          description: Failed to perform authorization over the entity.
+        "404":
+          description: A non-existent entity request.
+        "422":
+          description: Database can't process request.
+        "500":
+          $ref: "#/components/responses/ServiceError"
+
```
Comment (felixgateru) [inline api/openapi/users.yml]: same

File: pkg/sdk/go/users.go [hunk 23]
```diff
@@ -269,56 +269,73 @@ func (sdk mgSDK) UpdateUserRole(user User, token string) (User, errors.SDKError)
 	return user, nil
 }
 
-func (sdk mgSDK) ListUserChannels(userID string, pm PageMetadata, token string) (ChannelsPage, errors.SDKError) {
-	url, err := sdk.withQueryParams(sdk.thingsURL, fmt.Sprintf("%s/%s/%s", usersEndpoint, userID, channelsEndpoint), pm)
+func (sdk mgSDK) ListChannelUsers(pm PageMetadata, token string) (UsersPage, errors.SDKError) {
+	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/members", usersEndpoint), pm)
```
Comment (felixgateru) [inline pkg/sdk/go/users.go]:
```suggestion
	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s", usersEndpoint, membersEndpoint), pm)
```

File: pkg/sdk/go/users.go [hunk 24]
```diff
@@ -269,56 +269,73 @@ func (sdk mgSDK) UpdateUserRole(user User, token string) (User, errors.SDKError)
...
+	up := UsersPage{}
+	if err := json.Unmarshal(body, &up); err != nil {
+		return UsersPage{}, errors.NewSDKError(err)
 	}
 
+	return up, nil
+}
+
+func (sdk mgSDK) ListGroupUsers(pm PageMetadata, token string) (UsersPage, errors.SDKError) {
+	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/members", usersEndpoint), pm)
```
Comment (felixgateru) [inline pkg/sdk/go/users.go]: same

File: pkg/sdk/go/users.go [hunk 25]
```diff
@@ -269,56 +269,73 @@ func (sdk mgSDK) UpdateUserRole(user User, token string) (User, errors.SDKError)
...
 	}
 
-	return cp, nil
+	return up, nil
 }
 
-func (sdk mgSDK) ListUserGroups(userID string, pm PageMetadata, token string) (GroupsPage, errors.SDKError) {
-	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/%s/%s", usersEndpoint, userID, groupsEndpoint), pm)
+func (sdk mgSDK) ListThingUsers(pm PageMetadata, token string) (UsersPage, errors.SDKError) {
+	url, err := sdk.withQueryParams(sdk.usersURL, fmt.Sprintf("%s/members", usersEndpoint), pm)
```
Comment (felixgateru) [inline pkg/sdk/go/users.go]: same

File: api/openapi/things.yml [hunk 26]
```diff
@@ -962,6 +967,36 @@ paths:
         "500":
           $ref: "#/components/responses/ServiceError"
 
+  /channels/things:
```
Comment (felixgateru) [inline api/openapi/things.yml]: Does this endpoint exist?

File: api/openapi/things.yml [hunk 27]
```diff
@@ -401,7 +401,7 @@ paths:
         "500":
           $ref: "#/components/responses/ServiceError"
 
-  /channels/{chanID}/things:
+  /things/channels:
```
Comment (felixgateru) [inline api/openapi/things.yml]: Does this endpoint exist

**Review by felixgateru — CHANGES_REQUESTED** [review]

File: api/openapi/auth.yml [hunk 28]
```diff
@@ -427,39 +427,6 @@ paths:
           description: Missing or invalid content type.
         "500":
           $ref: "#/components/responses/ServiceError"
-  /users/{memberID}/domains:
```
Comment (felixgateru) [inline api/openapi/auth.yml:430]: Could you update `ListDomains` to reflect the addition of the userID query parameter

**Review by felixgateru — CHANGES_REQUESTED** [review]

**Review by felixgateru — DISMISSED** [review]
LGTM

**Comment by dborovcanin** [comment]
Merging this PR to the `api-refactor` branch so @arvindh123 can continue work from there.

**Pull request merged**